flink-table/flink-sql-client/src/main/java/org/apache/flink/util/ClientMutableURLClassLoader.java

```diff
@@ -0,0 +1,140 @@
+ */
+@Experimental
+@Internal
+public class ClientMutableURLClassLoader extends MutableURLClassLoader {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ClientMutableURLClassLoader.class);
+
+    static {
+        ClassLoader.registerAsParallelCapable();
+    }
+
+    private final Configuration configuration;
+    private final List<MutableURLClassLoader> oldClassLoaders = new ArrayList<>();
+    private MutableURLClassLoader currentClassLoader;
+
+    public ClientMutableURLClassLoader(
+            Configuration configuration, MutableURLClassLoader mutableURLClassLoader) {
+        super(new URL[0], mutableURLClassLoader);
+        this.configuration = configuration;
+        this.currentClassLoader = mutableURLClassLoader;
+    }
+
+    @Override
+    protected final Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        return currentClassLoader.loadClass(name, resolve);
+    }
+
+    @Override
+    public void addURL(URL url) {
+        currentClassLoader.addURL(url);
+    }
+
+    public void removeURL(URL url) {
+        Set<URL> registeredUrls =
+                Stream.of(currentClassLoader.getURLs()).collect(Collectors.toSet());
+        if (!registeredUrls.contains(url)) {
+            LOG.warn(
+                    String.format(
+                            "Could not remove the specified jar because the jar path [%s] is not found in classloader.",
+                            url));
+            return;
+        }
+
+        // add current classloader to list
+        oldClassLoaders.add(currentClassLoader);
+        // remove url from registeredUrls
+        registeredUrls.remove(url);
+        // update current classloader point to a new MutableURLClassLoader instance
+        currentClassLoader =
+                MutableURLClassLoader.newInstance(
+                        registeredUrls.toArray(new URL[0]),
+                        currentClassLoader.getParent(),
+                        configuration);
+    }
+
+    @Override
+    public URL[] getURLs() {
+        return currentClassLoader.getURLs();
+    }
+
+    @Override
+    public void close() throws IOException {
+        IOException exception = null;
+        try {
+            // close current classloader
+            currentClassLoader.close();
+        } catch (IOException e) {
+            LOG.debug("Error while closing class loader in ClientMutableURLClassLoader.", e);
+            exception = e;
+        }
+
+        // close other classloader in the list
+        for (MutableURLClassLoader classLoader : oldClassLoaders) {
+            try {
+                classLoader.close();
+            } catch (IOException ioe) {
+                LOG.debug(
+                        "Error while closing older class loader in ClientMutableURLClassLoader.",
+                        ioe);
+                exception = ExceptionUtils.firstOrSuppressed(ioe, exception);
```

Comment:
I think we should set the exception if it is null.

Reply:
What your means?